src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Validation.cs

```diff
@@ -753,5 +758,360 @@ private static ulong GetNonAsciiBytes(Vector128<byte> value, Vector128<byte> bit
+                    tooLong | overlong2 | twoConts | surrogate | tooLarge,
+                    tooLong | overlong2 | twoConts | surrogate | tooLarge,
+                    tooShort, tooShort, tooShort, tooShort, tooShort, tooShort, tooShort, tooShort,
+                    tooShort, tooShort, tooShort, tooShort,
+                    tooLong | overlong2 | twoConts | overlong3 | tooLarge1000 | overlong4,
+                    tooLong | overlong2 | twoConts | overlong3 | tooLarge,
+                    tooLong | overlong2 | twoConts | surrogate | tooLarge,
+                    tooLong | overlong2 | twoConts | surrogate | tooLarge,
+                    tooShort, tooShort, tooShort, tooShort);
+            Vector256<byte> thirdByte = Vector256.Create((byte)(0b11100000u - 0x80));
+            Vector256<byte> fourthByte = Vector256.Create((byte)(0b11110000u - 0x80));
+            Vector256<byte> v0f = Vector256.Create((byte)0x0F);
+            Vector256<byte> v80 = Vector256.Create((byte)0x80);
+            // So we want to count the number of 4-byte sequences,
+            // the number of 4-byte sequences, 3-byte sequences, and
+            // the number of 2-byte sequences.
+            // We can do it indirectly. We know how many bytes in total
+            // we have (length). Let us assume that the length covers
+            // only complete sequences (we need to adjust otherwise).
+            // We have that
+            // length = 4 * n4 + 3 * n3 + 2 * n2 + n1
+            // where n1 is the number of 1-byte sequences (ASCII),
+            // n2 is the number of 2-byte sequences, n3 is the number
+            // of 3-byte sequences, and n4 is the number of 4-byte sequences.
+            //
+            // Let ncon be the number of continuation bytes, then we have
+            // length =  n4 + n3 + n2 + ncon + n1
+            //
+            // We can solve for n2 and n3 in terms of the other variables:
+            // n3 = n1 - 2 * n4 + 2 * ncon - length
+            // n2 = -2 * n1 + n4 - 4 * ncon + 2 * length
+            // Thus we only need to count the number of continuation bytes,
+            // the number of ASCII bytes and the number of 4-byte sequences.
+            // But we need even less because we compute
+            // utfadjust = -2 * n4 - 2 * n3 - n2
+            // so n1 and length cancel out in the end. Thus we only need to compute
+            // n3' =  - 2 * n4 + 2 * ncon
+            // n2' = n4 - 4 * ncon
+            // The *block* here is what begins at processedLength and ends
+            // at processedLength/16*16 or when an error occurs.
+            // The block goes from processedLength to processedLength/16*16.
+            int countBytes = 0; // number of continuation bytes in the block
+            int n4 = 0; // number of 4-byte sequences that start in this block
+            for (; processedLength + 32 <= inputLength; processedLength += 32)
```

Comment:
```suggestion
            for (; processedLength <= inputLength - 32; processedLength += 32)
```

Reply:
It's not only better, but also more overflow-friendly. Sadly, I've just pushed a clean up change and these suggestions no longer applieable 🙂